python/tk_nuke/menu_generation.py

```diff
@@ -21,291 +18,665 @@
+            return
+
+        event.menu.addSeparator()
+        menu = event.menu.addAction("Shotgun")
+        menu.setEnabled(False)
+
+        for cmd in cmds:
+            enabled = True
+            if cmd.requires_selection:
+                if hasattr(event.sender, "selection") and not event.sender.selection():
+                    enabled = False
+            cmd.sender = event.sender
+            cmd.event_type = event.type
+            cmd.event_subtype = event.subtype
+            cmd.add_command_to_menu(event.menu, enabled)
+        event.menu.addSeparator()
+
+    def _add_context_menu(self):
+        """
+        Adds a context menu which displays the current context.
+        """
+        ctx = self.engine.context
+
+        if ctx.entity is None:
+            ctx_name = "%s" % ctx.project["name"]
+        elif ctx.step is None and ctx.task is None:
+            # entity only
+            # e.g. Shot ABC_123
+            ctx_name = "%s %s" % (ctx.entity["type"], ctx.entity["name"])
+        else:
+            # we have either step or task
+            task_step = None
+            if ctx.step:
+                task_step = ctx.step.get("name")
+            if ctx.task:
+                task_step = ctx.task.get("name")
+
+            # e.g. [Lighting, Shot ABC_123]
+            ctx_name = "%s, %s %s" % (task_step, ctx.entity["type"], ctx.entity["name"])
+
+        # create the menu object
+        ctx_menu = self._menu_handle.addMenu(ctx_name)
+        action = ctx_menu.addAction("Jump to Shotgun")
+        action.triggered.connect(self._jump_to_sg)
+        action = ctx_menu.addAction("Jump to File System")
+        action.triggered.connect(self._jump_to_fs)
+        ctx_menu.addSeparator()
+        return ctx_menu
+
+    def _add_app_menu(self, commands_by_app):
+        """
+        Add all apps to the main menu.
+        """
+        for app_name in sorted(commands_by_app.keys()):
+            if len(commands_by_app[app_name]) > 1:
+                # more than one menu entry fort his app
+                # make a sub menu and put all items in the sub menu
+                app_menu = self._menu_handle.addMenu(app_name)
+                for cmd in commands_by_app[app_name]:
+                    cmd.add_command_to_menu(app_menu)
+            else:
+                # this app only has a single entry.
+                # display that on the menu
+                # todo: Should this be labeled with the name of the app
+                # or the name of the menu item? Not sure.
+                cmd_obj = commands_by_app[app_name][0]
+                if not cmd_obj.favourite:
+                    # skip favourites since they are already on the menu
+                    cmd_obj.add_command_to_menu(self._menu_handle)
+
+# -----------------------------------------------------------------------------
+
+class NukeMenuGenerator(MenuGenerator):
+    """
+    A Nuke specific menu generator.
+    """
+    def __init__(self, *args, **kwargs):
```

Comment:
It makes little difference in this case, so I'm now passing the args explicitly. The advantage is that since the Nuke/Hiero specific __init__ methods don't require any arguments themselves, you can pass whatever is received up to the base class. If the __init__ args of the base class change in the future then there's no need to update its subclasses because those new args are passed up the chain automatically via args/kwargs.

In this case it's overkill since it's a very simple, self-contained class structure.